File: daq/slc/extra/javalib/src/b2daq/dqm/core/MonLabel.java
package b2daq.dqm.core;

import java.io.IOException;

import b2daq.core.Reader;
import b2daq.core.Writer;
import b2daq.graphics.HtmlColor;

public class MonLabel extends MonShape {

	private CharArray _array = new CharArray();
	private HtmlColor _font_color = new HtmlColor(HtmlColor.BLACK);
	
	public MonLabel(String name, int max_size) {
		super(name);
		_array.resize(max_size);
	}

	public MonLabel() {
		super();
		_array.resize(0);
	}

	@Override
	public String getDataType() {
		return "MLB";
	}
	
	public void writeConfig(Writer writer) throws IOException {
		super.writeConfig(writer);
		writer.writeInt(_array.length());
	}

	public void writeContents(Writer writer) throws IOException {
		super.writeContents(writer);
		writer.writeShort((short)_font_color.getRed());
		writer.writeShort((short)_font_color.getGreen());
		writer.writeShort((short)_font_color.getBlue());
		_array.writeObject(writer);
	}

	public void readConfig(Reader reader) throws IOException {
		super.readConfig(reader);
		_array.resize(reader.readInt());
	}

	public void readContents(Reader reader) throws IOException {
		super.readContents(reader);
		short red = reader.readShort();
		short green = reader.readShort();
		short blue = reader.readShort();
		_font_color.setRGB(red, green, blue);
		_array.readObject(reader);
	}

	public void setMaxSize(int max_size) {
		_array.resize(max_size);
	}

	public int getMaxSize() {
		return _array.length();
	}

	public void setText(String text) {
		int n = 0;
		for ( ; n < text.length() && n < _array.length() - 1; n++ ) {
			_array.set(n, text.charAt(n));
		}
		_array.set(n, '\n');
	}

	public String getText() {
		StringBuffer ss = new StringBuffer();
		for ( int n = 0; n < _array.length(); n++ ) {
			char c = _array.at(n);
			if ( c == '\n' || c == '\0' ) break;
			ss.append(c);
		}
		return ss.toString();
	}

	public void setFontColor(HtmlColor font_color) {
		_font_color = font_color;
	}

	public HtmlColor getFontColor() {
		return _font_color;
	}
	
	protected String getXML() {
		String str = "text=\"" + getText() + 
		"\" color=\"" + _font_color.toString() + 
		"\" >\n";
		return str;
	}

}

File: daq/slc/dqm/src/MonLabel.cc
#include "daq/slc/dqm/MonLabel.h"

#include <cstring>
#include <sstream>

using namespace Belle2;

MonLabel::MonLabel()
  : MonShape(), m_font_color(MonColor::BLACK), m_align("center") {}

MonLabel::MonLabel(const std::string& name, int max)
  : MonShape(name), m_font_color(MonColor::BLACK),
    m_text(max), m_align("center") {}

void MonLabel::setText(const std::string& text) throw()
{
  memset(m_text.ptr(), '\0', m_text.size());
  strncpy((char*)m_text.ptr(), text.c_str(), m_text.size());
}

std::string MonLabel::getText() const throw()
{
  return std::string((const char*)m_text.ptr());
}

void MonLabel::writeConfig(Writer& writer)
const throw(IOException)
{
  MonShape::writeConfig(writer);
  writer.writeInt(m_text.size());
}

void MonLabel::writeContents(Writer& writer)
const throw(IOException)
{
  MonShape::writeContents(writer);
  writer.writeShort((short)m_font_color.getRed());
  writer.writeShort((short)m_font_color.getGreen());
  writer.writeShort((short)m_font_color.getBlue());
  m_text.writeObject(writer);
}

void MonLabel::readConfig(Reader& reader)
throw(IOException)
{
  MonShape::readConfig(reader);
  m_text.resize(reader.readInt());
  m_text.setAll('\0');
}

void MonLabel::readContents(Reader& reader)
throw(IOException)
{
  MonShape::readContents(reader);
  short red = reader.readShort();
  short green = reader.readShort();
  short blue = reader.readShort();
  m_font_color.setRGB(red, green, blue);
  m_text.readObject(reader);
}

std::string MonLabel::toString() const throw()
{
  std::stringstream ss;
  ss << "<mon-object data-type='" << getDataType() << "' "
     << "name='" << m_name << "' "
     << "update-id='" << m_update_id << "' "
     << "tab='" << (int) m_tab_id << "' "
     << "position='" << (int) m_position_id << "' "
     << "text='" << getText() << "' />";
  return ss.str();
}

std::string MonLabel::toXML() const throw()
{
  std::stringstream ss;
  ss << "<" << ELEMENT_TAG << " "
     << "name='" << m_name << "' ";
  if (m_line_pro != NULL) ss << m_line_pro->toString();
  if (m_fill_pro != NULL) ss << m_fill_pro->toString();
  if (m_font_pro != NULL) ss << m_font_pro->toString();
  if (m_x >= 0 && m_y >= 0 && m_width >= 0 && m_height >= 0) {
    ss << "bounds='" << m_x << "," << m_y << ","
       << m_width << "," << m_height << "' ";
  }
  ss << "text-align='" << m_align << "' "
     << "/>\n";
  return ss.str();
}

